File: src/utils/geometry.py
from PyQt6.QtCore import QByteArray, QSettings
from PyQt6.QtWidgets import QWidget

class GeometryManager:
    def __init__(self, window: QWidget, settings: QSettings, store=None):
        self.window = window
        self.settings = settings
        self.store = store

        self.normal_geometry: QByteArray | None = None
        self.normal_rect_str: str | None = None
        self._freeze_normal_updates: bool = False

    def load_and_apply(self):

        if self.store and hasattr(self.store.settings, 'window_width'):
            was_maximized = self.store.settings.window_was_maximized
            window_width = max(200, self.store.settings.window_width)
            window_height = max(150, self.store.settings.window_height)
            window_x = self.store.settings.window_x
            window_y = self.store.settings.window_y

            self.window.setGeometry(window_x, window_y, window_width, window_height)
            self.normal_rect_str = f"{window_x},{window_y},{window_width},{window_height}"
        else:

            was_maximized = self.settings.value("window_was_maximized", False, type=bool)
            saved_normal_geom = self.settings.value("normal_geometry", QByteArray(), type=QByteArray)
            saved_normal_rect_str = self.settings.value("normal_rect", "", type=str)

            if saved_normal_rect_str:
                try:
                    parts = [int(p) for p in saved_normal_rect_str.split(",")]
                    if len(parts) == 4:
                        x, y, w, h = parts
                        self.window.setGeometry(x, y, max(200, w), max(150, h))
                        self.normal_rect_str = saved_normal_rect_str
                except Exception:
                    pass

            if not self.normal_rect_str:
                if saved_normal_geom and not saved_normal_geom.isNull():
                    self.normal_geometry = saved_normal_geom
                    self.window.restoreGeometry(self.normal_geometry)
                else:
                    self.window.setGeometry(100, 100, 1024, 768)
                    self.normal_geometry = self.window.saveGeometry()

        self._freeze_normal_updates = bool(was_maximized)

        if was_maximized:
            self.window.showMaximized()
        else:
            self.window.showNormal()

    def update_normal_geometry_if_needed(self):
        if self._freeze_normal_updates:
            return
        if not self.window.isMaximized() and not self.window.isFullScreen():
            try:
                self.normal_geometry = self.window.saveGeometry()
            except Exception:
                pass
            try:
                g = self.window.geometry()
                self.normal_rect_str = f"{g.x()},{g.y()},{g.width()},{g.height()}"
            except Exception:
                pass

    def save_on_close(self):
        is_maximized = self.window.isMaximized() or self.window.isFullScreen()

        if self.store and hasattr(self.store.settings, 'window_width'):
            self.store.settings.window_was_maximized = is_maximized

            if not is_maximized and self.normal_rect_str:
                try:
                    parts = [int(p) for p in self.normal_rect_str.split(",")]
                    if len(parts) == 4:
                        x, y, w, h = parts
                        self.store.settings.window_x = x
                        self.store.settings.window_y = y
                        self.store.settings.window_width = w
                        self.store.settings.window_height = h
                except Exception:
                    pass

        self.settings.setValue("window_was_maximized", is_maximized)

        if self.normal_geometry and not self.normal_geometry.isNull():
            self.settings.setValue("normal_geometry", self.normal_geometry)
        if self.normal_rect_str:
            self.settings.setValue("normal_rect", self.normal_rect_str)

        self.settings.sync()

    def on_left_maximized_state(self):
        self._freeze_normal_updates = False
        self.update_normal_geometry_if_needed()

    def begin_maximize_transition(self):
        try:
            g = self.window.geometry()
            self.normal_rect_str = f"{g.x()},{g.y()},{g.width()},{g.height()}"
        except Exception:
            pass
        self._freeze_normal_updates = True

File: cpp/app/utils/geometry.cpp
#include "utils/geometry.h"

#include <QStringList>

#include <algorithm>

namespace imgsli::app::utils {

namespace {

constexpr int kMinWidth = 200;
constexpr int kMinHeight = 150;

}  // namespace

GeometryManager::GeometryManager(QWidget* window, QSettings* settings,
                                 Store* store, QObject* parent)
    : QObject(parent), window_(window), settings_(settings), store_(store) {}

void GeometryManager::loadAndApply() {
  if (window_ == nullptr || settings_ == nullptr) {
    return;
  }

  const bool wasMaximized =
      settings_->value(QStringLiteral("window_was_maximized"), false).toBool();
  const QByteArray savedNormalGeom =
      settings_->value(QStringLiteral("normal_geometry"), QByteArray())
          .toByteArray();
  const QString savedNormalRectStr =
      settings_->value(QStringLiteral("normal_rect"), QString()).toString();

  bool restored = false;
  if (!savedNormalRectStr.isEmpty()) {
    const QStringList parts = savedNormalRectStr.split(QLatin1Char(','));
    if (parts.size() == 4) {
      bool ok = true;
      const int x = parts[0].toInt(&ok);
      const int y = parts[1].toInt(&ok);
      const int w = parts[2].toInt(&ok);
      const int h = parts[3].toInt(&ok);
      if (ok) {
        window_->setGeometry(x, y, std::max(kMinWidth, w),
                              std::max(kMinHeight, h));
        normalRectStr_ = savedNormalRectStr;
        restored = true;
      }
    }
  }

  if (!restored) {
    if (!savedNormalGeom.isNull()) {
      normalGeometry_ = savedNormalGeom;
      window_->restoreGeometry(savedNormalGeom);
    } else {
      window_->setGeometry(100, 100, 1024, 768);
      normalGeometry_ = window_->saveGeometry();
    }
  }

  freezeNormalUpdates_ = wasMaximized;

  const Qt::WindowStates state = window_->windowState();
  if (wasMaximized) {
    window_->setWindowState(state | Qt::WindowMaximized);
  } else {
    window_->setWindowState(state & ~Qt::WindowMaximized);
  }
}

void GeometryManager::updateNormalGeometryIfNeeded() {
  if (freezeNormalUpdates_ || window_ == nullptr) {
    return;
  }
  if (!window_->isMaximized() && !window_->isFullScreen()) {
    normalGeometry_ = window_->saveGeometry();
    const QRect g = window_->geometry();
    normalRectStr_ = QStringLiteral("%1,%2,%3,%4")
                          .arg(g.x())
                          .arg(g.y())
                          .arg(g.width())
                          .arg(g.height());
  }
}

void GeometryManager::saveOnClose() {
  if (window_ == nullptr || settings_ == nullptr) {
    return;
  }
  const bool isMaximized = window_->isMaximized() || window_->isFullScreen();
  settings_->setValue(QStringLiteral("window_was_maximized"), isMaximized);
  if (normalGeometry_.has_value() && !normalGeometry_->isNull()) {
    settings_->setValue(QStringLiteral("normal_geometry"), *normalGeometry_);
  }
  if (normalRectStr_.has_value() && !normalRectStr_->isEmpty()) {
    settings_->setValue(QStringLiteral("normal_rect"), *normalRectStr_);
  }
  settings_->sync();
}

void GeometryManager::onLeftMaximizedState() {
  freezeNormalUpdates_ = false;
  updateNormalGeometryIfNeeded();
}

void GeometryManager::beginMaximizeTransition() {
  if (window_ == nullptr) {
    return;
  }
  const QRect g = window_->geometry();
  normalRectStr_ = QStringLiteral("%1,%2,%3,%4")
                        .arg(g.x())
                        .arg(g.y())
                        .arg(g.width())
                        .arg(g.height());
  freezeNormalUpdates_ = true;
}

}  // namespace imgsli::app::utils
